server/src/main/java/org/opensearch/index/translog/RemoteFsTranslog.java

```diff
@@ -0,0 +1,313 @@
+            }
+            Collections.reverse(foundTranslogs);
+
+            // when we clean up files, we first update the checkpoint with a new minReferencedTranslog and then delete them;
+            // if we crash just at the wrong moment, it may be that we leave one unreferenced file behind so we delete it if there
+            IOUtils.deleteFilesIgnoringExceptions(
+                location.resolve(Translog.getFilename(minGenerationToRecoverFrom - 1)),
+                location.resolve(Translog.getCommitCheckpointFileName(minGenerationToRecoverFrom - 1))
+            );
+
+            Path commitCheckpoint = location.resolve(Translog.getCommitCheckpointFileName(checkpoint.generation));
+            if (Files.exists(commitCheckpoint)) {
+                Checkpoint checkpointFromDisk = Checkpoint.read(commitCheckpoint);
+                if (checkpoint.equals(checkpointFromDisk) == false) {
+                    throw new TranslogCorruptedException(
+                        commitCheckpoint.toString(),
+                        "checkpoint file "
+                            + commitCheckpoint.getFileName()
+                            + " already exists but has corrupted content: expected "
+                            + checkpoint
+                            + " but got "
+                            + checkpointFromDisk
+                    );
+                }
+            } else {
+                copyCheckpointTo(commitCheckpoint);
+            }
+            success = true;
+        } finally {
+            if (success == false) {
+                IOUtils.closeWhileHandlingException(foundTranslogs);
+            }
+        }
+        return foundTranslogs;
+    }
+
+    @Override
+    boolean ensureSynced(Location location) throws IOException {
+        Callable<Boolean> execute = () -> false;
+        try (ReleasableLock lock = readLock.acquire()) {
+            assert location.generation <= current.getGeneration();
+            if (location.generation == current.getGeneration()) {
+                ensureOpen();
+                execute = () -> prepareAndUpload(primaryTermSupplier.getAsLong(), location.generation);
+            }
+        } catch (final Exception ex) {
+            closeOnTragicEvent(ex);
+            throw ex;
+        }
+        try {
+            return execute.call();
+        } catch (Exception ex) {
+            closeOnTragicEvent(ex);
+            assert ex instanceof IOException;
+            throw (IOException) ex;
+        }
+    }
+
+    @Override
+    public void rollGeneration() throws IOException {
+        syncBeforeRollGeneration();
+        if (current.totalOperations() == 0 && primaryTermSupplier.getAsLong() == current.getPrimaryTerm()) {
+            return;
+        }
+        prepareAndUpload(primaryTermSupplier.getAsLong(), null);
+    }
+
+    private boolean prepareAndUpload(Long primaryTerm, Long generation) throws IOException {
```

Comment:
Does `prepareAndUpload` upload only the latest sync'd tarnslog and checkpoint files? Does it assume existence of previous files in remote translog?

Reply:
No it doesn't . It uploads the difference b/w local and remote stores.  The `transferSnapshotProvider` factors in all the local translog and checkpoint files and uses `FileTransferTracker` to know the already uploaded files.